sound/soc/sof/intel/hda-dai.c

```diff
@@ -151,49 +151,77 @@ static int hda_link_dma_params(struct hdac_ext_stream *stream,
 	return 0;
 }
 
-/* Send DAI_CONFIG IPC to the DAI that matches the dai_name and direction */
-static int hda_link_config_ipc(struct sof_intel_hda_stream *hda_stream,
-			       const char *dai_name, int channel, int dir)
+/* find the snd_sof_dai matching the BE CPU DAI name and update its config */
+static struct snd_sof_dai *hda_get_sof_dai(struct sof_intel_hda_stream *hda_stream,
+					   const char *dai_name, int channel, int dir)
 {
+	struct snd_sof_dev *sdev = hda_stream->sdev;
 	struct sof_ipc_dai_config *config;
 	struct snd_sof_dai *sof_dai;
-	struct sof_ipc_reply reply;
-	int ret = 0;
+	bool found = false;
 
-	list_for_each_entry(sof_dai, &hda_stream->sdev->dai_list, list) {
+	/* find snd_sof_dai matching the BE CPU DAI name */
+	list_for_each_entry(sof_dai, &sdev->dai_list, list) {
 		if (!sof_dai->cpu_dai_name)
 			continue;
 
 		if (!strcmp(dai_name, sof_dai->cpu_dai_name) &&
 		    dir == sof_dai->comp_dai.direction) {
-			config = sof_dai->dai_config;
+			found = true;
+			break;
+		}
+	}
 
-			if (!config) {
-				dev_err(hda_stream->sdev->dev,
-					"error: no config for DAI %s\n",
-					sof_dai->name);
-				return -EINVAL;
-			}
+	if (!found) {
+		dev_err(sdev->dev, "error: failed to find SOF DAI for: %s\n", dai_name);
+		return NULL;
+	}
 
-			/* update config with stream tag */
-			config->hda.link_dma_ch = channel;
+	config = sof_dai->dai_config;
+	if (!config) {
+		dev_err(sdev->dev, "error: no config for DAI %s\n", sof_dai->name);
+		return NULL;
+	}
 
-			/* send IPC */
-			ret = sof_ipc_tx_message(hda_stream->sdev->ipc,
-						 config->hdr.cmd,
-						 config,
-						 config->hdr.size,
-						 &reply, sizeof(reply));
+	/* update config with stream tag */
+	config->hda.link_dma_ch = channel;
 
-			if (ret < 0)
-				dev_err(hda_stream->sdev->dev,
-					"error: failed to set dai config for %s\n",
-					sof_dai->name);
-			return ret;
-		}
-	}
+	return sof_dai;
+}
+
+static int hda_link_config_ipc(struct sof_intel_hda_stream *hda_stream,
+			       const char *dai_name, int channel, int dir)
+{
+	struct snd_sof_dev *sdev = hda_stream->sdev;
+	struct sof_ipc_dai_config *config;
+	struct snd_sof_dai *sof_dai;
+	struct sof_ipc_reply reply;
+
+	sof_dai = hda_get_sof_dai(hda_stream, dai_name, channel, dir);
+	if (IS_ERR(sof_dai))
+		return PTR_ERR(sof_dai);
```

Comment:
You never return `ERR_PTR()` from `hda_get_sof_dai()` you return `NULL` on error.

Reply:
yes, thanks. I can never get the ERR_PTR()/PTR_ERR() bits right!